### PR TITLE
Skip DirtyHandler when RouterTabs are used beneath a Form

### PR DESCRIPTION
**Pull request by fraxachun**
Add a somewhat fancy mechanism to store the parentFormId in the Location-Object
which is then used to skip the isDirty-Check in a FinalForm

https://user-images.githubusercontent.com/1013756/142583291-e99a6c33-cb18-4073-9c18-f970a9c8cfb1.mp4

### CODE REVIEW

File: packages/admin/src/FinalForm.tsx [hunk 1]
```diff
@@ -12,6 +14,7 @@ import { FinalFormContext, FinalFormContextProvider } from "./form/FinalFormCont
 import { SubmitError, SubmitResult } from "./form/SubmitResult";
 import { StackApiContext } from "./stack/Api";
 import { TableQueryContext } from "./table/TableQueryContext";
+const UUID = require("uuid");
```
Comment (kaufmo) [inline packages/admin/src/FinalForm.tsx:17]: why do you use require here?

Comment (fraxachun) [reply packages/admin/src/FinalForm.tsx:17]: Updated in this PR: https://github.com/vivid-planet/comet-admin/pull/546

File: packages/admin/src/tabs/RouterTabs.tsx [hunk 2]
```diff
@@ -33,12 +34,15 @@ function RouterTabsComponent({
...
     const childrenArr = React.Children.toArray(children);
+    const finalFormContext = useFinalFormContext();
 
     const handleChange = (event: {}, value: number) => {
         const paths = childrenArr.map((child) => {
             return React.isValidElement<TabProps>(child) ? child.props.path : null;
         });
-        history.push(match.url + paths[value]);
+        history.push(match.url + paths[value], {
+            parentFormId: finalFormContext.formId,
```
Comment (nsams) [inline packages/admin/src/tabs/RouterTabs.tsx:44]: This state is not there when eg. copy/pasting this url in a new browser window. What happens then?

File: packages/admin/src/DirtyHandlerApiContext.tsx [hunk 3]
```diff
@@ -1,9 +1,10 @@
+import { Location } from "history";
 import * as React from "react";
 
 import { SubmitResult } from "./form/SubmitResult";
 
 export interface IDirtyHandlerApiBinding {
-    isDirty: () => boolean;
+    isDirty: (location?: Location) => boolean;
```
Comment (nsams) [inline packages/admin/src/DirtyHandlerApiContext.tsx:7]: I don't understand how this works :)

what location is passed here? can you name it more specific?

Comment (fraxachun) [reply packages/admin/src/DirtyHandlerApiContext.tsx]: You're right, it's not logical to pass location into isDirty. Have a look if you like it this way or whether we should type the payload for isDirty.

File: packages/admin/src/tabs/RouterTabs.tsx (hunk 2 above)
Comment (fraxachun) [reply packages/admin/src/tabs/RouterTabs.tsx:44]: I don't understand the problem. The state is completely optional, if it's not set nothing happens.

**Review by johnnyomair — APPROVED** [review]

**Comment by fraxachun** [comment]
> I imagined something else, like a container (that FinalForm adds) that says "dont do dirty checking in subroutes"
I can't find a clean solution. The RouterTabs-Component pushes directory into the react-router history (https://github.com/vivid-planet/comet-admin/pull/545/files#diff-fbda3187bd3ce6fdc4f090e13b52daedf2e321de10ffb0931cbae4870d86d5a0R43) and this is the only possibility to store the context about the current position in the DOM-Tree. However, I could create a new context HistoryState which is created by FinalForm and used by RouterTabs to store basically the same data like now in the state. Maybe it makes it a bit cleaner, but it's the same logic like now. That's all I can think of.

**Comment by nsams** [comment]
react-router Prompt gets passed in the new location (see https://v5.reactrouter.com/core/api/Prompt), that could be used to not show the message in sub routes

**Comment by thomasdax98** [comment]
When navigating from tab 1 to tab 2, all fields in tab 1 are unregistered. Hence, the forms `pristine` and `dirty` state is lost.

This is problematic because on many occasions we use something like this for the save button:

```tsx
<Form>
  {({ pristine, hasValidationErrors, submitting, hasSubmitErrors, handleSubmit }) => {
    return <>
      <SplitButton
            disabled={pristine || hasValidationErrors || submitting}
            localStorageKey={"routertabs-with-forms-save"}
        >
            <SaveButton
                color={"primary"}
                variant={"contained"}
                saving={submitting}
                hasErrors={hasSubmitErrors}
                type="button"
                onClick={async () => {
                    handleSubmit();
                }}
            >
                Save
            </SaveButton>
            <SaveButton
                color={"primary"}
                variant={"contained"}
                saving={submitting}
                hasErrors={hasSubmitErrors}
                onClick={async () => {
                    handleSubmit();
                }}
            >
                Save and go back
            </SaveButton>
        </SplitButton>
       {/* ... other stuff ... */}
    </>
  }}
</Form>
```

The button is disabled if the form is `pristine`. Therefore, when navigating from tab 1 to tab 2, the button is disabled again although there a value in the form has changed (see video below).

https://user-images.githubusercontent.com/13380047/143838000-ba280cca-b1e4-477c-8a98-d261c25b1c66.mov

Demo here: https://deploy-preview-551--comet-admin.netlify.app/?path=/story/comet-admin-form--routertabs-with-forms

Do you have an idea how to handle this?

**Comment by fraxachun** [comment]
> react-router Prompt gets passed in the new location (see https://v5.reactrouter.com/core/api/Prompt), that could be used to not show the message in sub routes

If you have a look at the story (https://github.com/vivid-planet/comet-admin/blob/531e3a8a3ee806e3b38d063a37e82f2b7304f7e7/packages/admin-stories/src/admin/form/RouterTabs.tsx) you can see that the main panel has no route, even though this structure already exists: DirtyHandler -> RouterTabs -> FinalForm -> RouterTabs. So in my opinion it is not possible to decide if a location change occurs under a FinalForm-Component without having the source of the change (RouterTabs) adding context information.

**Comment by dkarnutsch** [comment]
@nsams can please have another look at this?

**Pull request closed without merging**